### ISSUE TITLE
In case processes are still running when trying to close a session, say which ones

### ISSUE THREAD

**Issue by GoogleCodeExporter** [mintty/mintty#448]
```
Clicking on the "X" in the window's title bar may result in:

---------------------------
mintty
---------------------------
Processes are running in session.
Close anyway?
---------------------------
OK   Cancel   
---------------------------

It would be nice to actually list as part of the message which processes these 
are.
```

Original issue reported on code.google.com by `sschuberth` on 14 Apr 2015 at 7:33

**Comment by mintty** [comment]
Please try the child process display feature (use the "Download ZIP" button for a test version).

**Comment by sschuberth** [comment]
I've tried it, but it does not seem to work correctly. I've replaced the mintty.exe from Git for Windows 2.x with one built from master, started Git Bash, launched `gitk &`, and then clicked on the window's close button. The prompt I was getting just shows an empty list:

```
---------------------------
mintty
---------------------------
Processes are running in session:
Close anyway?
---------------------------
OK   Cancel   
---------------------------
```

**Comment by mintty** [comment]
I forgot to point out that this feature requires procps to be installed. I don't know what specific adapted cygwin environment Git for Windows brings along but that's probably the missing issue.
Maybe I'll change the solution to use ps rather than procps.

**Comment by sschuberth** [comment]
Git for Windows 2.x is based on MSYS2, not on Cygwin directly. However, luckily [MSYS2 has a package for procps](https://github.com/Alexpux/MSYS2-packages/tree/master/procps), so I'll give it a try again with having that installed.

**Comment by sschuberth** [comment]
So I've installed `procps` from the MSYS2 repository to my Git for Windows 2.x directory and it basically seems to be working now. I get

    ---------------------------
    mintty
    ---------------------------
    Processes are running in session:
      PID TTY          TIME CMD
     1376 pty0     00:00:00 bash
    10756 pty0     00:00:02 wish
    Close anyway?
    ---------------------------
    OK   Cancel   
    ---------------------------

The only odd thing is that the reported PIDs do not match the ones reported by Windows task manager for these processes. Compared to the output of `ps`, it seems that `procps` only reports the `PID`, but not the `WINPID`. Moreover, there seems to be no way for `procps` to (also) output the `WINPID`. As IMO the `WINPID` would make more sense here, maybe indeed use `ps` instead of `procps`?

**Comment by mintty** [comment]
Unfortunately, cygwin ps is a limited implementation of ps and not very configurable. There is only one output format including the WINPID which clutters the output with useless overhead and also it's too long for the message box so already the header line wraps, like this:
```
---------------------------
Processes are running in session:
      PID    PPID    PGID     WINPID   TTY     UID    STIME 
COMMAND
     9076    5596    9076       7680  pty3    133715 13:41:37 
/usr/bin/bash
     1924    9076    1924       9084  pty3    133715 13:41:54 
/usr/bin/sleep
Close anyway?
---------------------------
```
I tried to filter the output with `cut` but that doesn't work because the STIME column may contain a space. Of course I could filter it myself... I can also make the `ps` command a configurable preference.

**Comment by sschuberth** [comment]
I see. In that case `procps`'s output is certainly better than nothing as it gives the process name so you'll like be able to locate and close it. Thanks for implementing this feature!

**Comment by mintty** [comment]
I like closing issues :)
Current approach: procps -o pid,ruser=USER -o comm 2> /dev/null || ps -ef
in order to work whether procps is installed or not; to be fed into popen().
I assume then that you do not wish a configurable ps command.
Thanks for discussing the details.

**Comment by mintty** [comment]
Uploaded to repository: the process list for exit confirmation is now created from the /proc filesystem, without starting an external `ps` process.
This speeds up the information pop-up and likely addresses https://cygwin.com/ml/cygwin/2018-10/msg00058.html in case of aggressive virus scanners.
It also adds the opportunity to include the WINPID in the listing.

**Comment by mintty** [comment]
Released 2.9.4; do spawn an external process (`ps`) anymore, display of process list is faster and blocking in case of aggressive virus scanners is avoided.
